### ISSUE TITLE
undefined method `next_token' for nil:NilClass

### ISSUE THREAD

**Issue by bitfield** [rodjek/puppet-lint#824]
puppet-lint version: 2.3.5
ruby version: 2.5.1-p57
platform: x86_64-darwin15
file path: modules/phpbitfield/manifests/init.pp
file contents:
```
# Manage PHP, old-school style
class phpbitfield {
  require repo::ius
  require repo::webtatic
  include phpbitfield::composer

  $php = lookup('phpbitfield::version', String[1])
  if $php == 'php56w' {
    $apc = 'apcu' # php56w-pecl-apc currently unavailable in webtatic repo
  } else {
    $apc = 'apc'
  }

  if $manage_repos {
    class { '::php::repo': }
    -> Anchor['php::begin']
  }

  anchor { 'php::begin': }
    -> class { '::php::packages': }
    -> class { '::php::cli':
      settings => $real_settings,
    }
  -> anchor { 'php::end': }

  # Configure global PHP settings in php.ini
  if $facts['os']['family'] != 'Debian' {
    Class['php::packages']
    -> class {'::php::global':
      settings => $real_settings,
    }
    -> Anchor['php::end']
  }

  file { '/etc/php.ini':
    content => epp('phpbitfield/php.ini.epp',
      {
        'display_errors'      => lookup('phpbitfield::display_errors'),
        'post_max_size'       => lookup('phpbitfield::post_max_size'),
        'upload_max_filesize' => lookup('phpbitfield::upload_max_filesize'),
        'memory_limit'        => lookup('phpbitfield::memory_limit'),
      }
    -> Anchor['php::end']
  }

  file { '/etc/php.d/mcrypt.ini':
    source  => 'puppet:///modules/phpbitfield/mcrypt.ini',
    require => Package["${php}-common"],
  }

  file { '/etc/php.d/ldap.ini':
    source  => 'puppet:///modules/phpbitfield/ldap.ini',
    require => Package["${php}-common"],
  }
}

```
error:
```
NoMethodError: undefined method `next_token' for nil:NilClass
/usr/local/lib/ruby/gems/2.5.0/gems/puppet-lint-2.3.5/lib/puppet-lint/lexer/token.rb:196:in `find_token_of'
/usr/local/lib/ruby/gems/2.5.0/gems/puppet-lint-2.3.5/lib/puppet-lint/lexer/token.rb:142:in `next_token_of'
/usr/local/lib/ruby/gems/2.5.0/gems/puppet-lint-2.3.5/lib/puppet-lint/data.rb:176:in `block in resource_indexes'
/usr/local/lib/ruby/gems/2.5.0/gems/puppet-lint-2.3.5/lib/puppet-lint/data.rb:171:in `each'
/usr/local/lib/ruby/gems/2.5.0/gems/puppet-lint-2.3.5/lib/puppet-lint/data.rb:171:in `resource_indexes'
/usr/local/lib/ruby/gems/2.5.0/gems/puppet-lint-2.3.5/lib/puppet-lint/checkplugin.rb:81:in `resource_indexes'
/usr/local/lib/ruby/gems/2.5.0/gems/puppet-lint-2.3.5/lib/puppet-lint/plugins/check_conditionals/selector_inside_resource.rb:7:in `check'
/usr/local/lib/ruby/gems/2.5.0/gems/puppet-lint-2.3.5/lib/puppet-lint/checkplugin.rb:21:in `run'
/usr/local/lib/ruby/gems/2.5.0/gems/puppet-lint-2.3.5/lib/puppet-lint/checks.rb:60:in `block in run'
/usr/local/lib/ruby/gems/2.5.0/gems/puppet-lint-2.3.5/lib/puppet-lint/checks.rb:57:in `each'
/usr/local/lib/ruby/gems/2.5.0/gems/puppet-lint-2.3.5/lib/puppet-lint/checks.rb:57:in `run'
/usr/local/lib/ruby/gems/2.5.0/gems/puppet-lint-2.3.5/lib/puppet-lint.rb:205:in `run'
/Users/john/git/thirtythree/33-puppet/Rakefile:100:in `block (3 levels) in <top (required)>'
/Users/john/git/thirtythree/33-puppet/Rakefile:98:in `each'
/Users/john/git/thirtythree/33-puppet/Rakefile:98:in `block (2 levels) in <top (required)>'
/usr/local/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/file_utils_ext.rb:59:in `verbose'
/Users/john/git/thirtythree/33-puppet/Rakefile:90:in `block in <top (required)>'
/usr/local/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:251:in `block in execute'
/usr/local/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:251:in `each'
/usr/local/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:251:in `execute'
/usr/local/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:195:in `block in invoke_with_call_chain'
/usr/local/Cellar/ruby/2.5.1/lib/ruby/2.5.0/monitor.rb:226:in `mon_synchronize'
/usr/local/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:188:in `invoke_with_call_chain'
/usr/local/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/task.rb:181:in `invoke'
/usr/local/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:160:in `invoke_task'
/usr/local/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:116:in `block (2 levels) in top_level'
/usr/local/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:116:in `each'
/usr/local/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:116:in `block in top_level'
/usr/local/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:125:in `run_with_threads'
/usr/local/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:110:in `top_level'
/usr/local/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:83:in `block in run'
/usr/local/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:186:in `standard_exception_handling'
/usr/local/lib/ruby/gems/2.5.0/gems/rake-12.3.0/lib/rake/application.rb:80:in `run'
/usr/local/lib/ruby/gems/2.5.0/gems/rake-12.3.0/exe/rake:27:in `<top (required)>'
/usr/local/bin/rake:23:in `load'
/usr/local/bin/rake:23:in `<main>'
```

**Comment by PeterDaveHello** [comment]
+1

**Comment by PeterDaveHello** [comment]
The interesting part: the only only appears when using `--fix` parameter, without it, it works without any issue, recognizes indentation issues properly.

**Comment by abegosum** [comment]
Seeing a similar issue.

puppet-lint version: 2.3.5
ruby version: 2.3.1-p112
platform: x86_64-linux-gnu
file path: ./manifests/accounts/mpr_app_user.pp
file contents:
```
#defined type for application users
define profiles::accounts::mpr_app_user (
  $uid,
  $gid,
  $groups,
  $realname,
  $pass,
  $ensure = 'present',
  $shell = '/bin/bash',
  $create_home = false,
  $username = $title,
  $home = "/export/app_users/${title}",
  $expiry = false,
  $purge_ssh_keys = false,
) {

  $is_dmz = hiera('is_dmz', str2bool($::is_dmz))
  $home_is_automounted = $::automount_running

  if (!$expiry)
  {
    user { $title:
      ensure         => $ensure,
      uid            => $uid,
      gid            => $gid,
      groups         => $groups,
      shell          => $shell,
      home           => $home,
      comment        => $realname,
      password       => $pass,
      managehome     => false,
      name           => $username,
      purge_ssh_keys => $purge_ssh_keys,
    }
  } else {
    user { $title:
      ensure         => $ensure,
      uid            => $uid,
      gid            => $gid,
      groups         => $groups,
      shell          => $shell,
      home           => $home,
      comment        => $realname,
      password       => $pass,
      managehome     => false,
      name           => $username,
      expiry         => $expiry,
      purge_ssh_keys => $purge_ssh_keys,
    }
  }

  if ($create_home) {
    if ($ensure == 'present' and !($home_is_automounted)) {
      exec { "${title} homedir":
        command => "mkdir -p ${home}; chown ${uid}:${gid} ${home}",
        path    => '/bin:/usr/sbin',
        creates => $home,
        require => User[$title]
      }
      ->
      file { $home :
        ensure => 'directory',
        owner  => $uid,
        group  => $gid,
        mode   => '0755',
      }
    } elsif ($ensure == 'absent') {
      file { $home :
        ensure => 'absent',
      }
    }
  }
}

```
error:
```
NoMethodError: undefined method `next_token_of' for nil:NilClass
/var/lib/gems/2.3.0/gems/puppet-lint-2.3.5/lib/puppet-lint/plugins/check_resources/ensure_first_param.rb:32:in `fix'
/var/lib/gems/2.3.0/gems/puppet-lint-2.3.5/lib/puppet-lint/checkplugin.rb:42:in `block in fix_problems'
/var/lib/gems/2.3.0/gems/puppet-lint-2.3.5/lib/puppet-lint/checkplugin.rb:38:in `each'
/var/lib/gems/2.3.0/gems/puppet-lint-2.3.5/lib/puppet-lint/checkplugin.rb:38:in `fix_problems'
/var/lib/gems/2.3.0/gems/puppet-lint-2.3.5/lib/puppet-lint/checks.rb:63:in `block in run'
/var/lib/gems/2.3.0/gems/puppet-lint-2.3.5/lib/puppet-lint/checks.rb:57:in `each'
/var/lib/gems/2.3.0/gems/puppet-lint-2.3.5/lib/puppet-lint/checks.rb:57:in `run'
/var/lib/gems/2.3.0/gems/puppet-lint-2.3.5/lib/puppet-lint.rb:205:in `run'
/var/lib/gems/2.3.0/gems/puppet-lint-2.3.5/lib/puppet-lint/bin.rb:66:in `block in run'
/var/lib/gems/2.3.0/gems/puppet-lint-2.3.5/lib/puppet-lint/bin.rb:62:in `each'
/var/lib/gems/2.3.0/gems/puppet-lint-2.3.5/lib/puppet-lint/bin.rb:62:in `run'
/var/lib/gems/2.3.0/gems/puppet-lint-2.3.5/bin/puppet-lint:7:in `<top (required)>'
/usr/local/bin/puppet-lint:23:in `load'
/usr/local/bin/puppet-lint:23:in `<main>'
```

**Comment by TJM** [comment]
Also seeing a similar issue here... only when using `-f` (or --fix)

Without the --fix, I am getting
```
base_aix/packages.pp - WARNING: ensure found on line but it's not the first attribute on line 54
```

or specifically:
```
      attributes => { ensure => 'mounted' },
```

Changing it to:
```
      attributes => {
        ensure => 'mounted',
      },
```
... doesn't help.

However, changing it to:
```
      attributes => {ensure => 'mounted' }, # lint:ignore:ensure_first_param
```
works.


Here is the rest of the requested output:
puppet-lint version: 2.3.6
ruby version: 2.4.3-p205
platform: x86_64-darwin16
file path: ./base_aix/packages.pp
file contents:
```
# AIX Base Packages
class profile::base_aix::packages (
  Optional[String] $repo_mount = undef,
  String $repo_path = '/var/run/pkg_repo',
) {

  # Once yum_bundle is approved and setup, we will change this so that it retrieves from yum.
  package { 'bash':
    ensure   => 'present',
    provider => 'rpm',
    source   => '/share/pkg/rpm/bash-4.4-3.aix6.1.ppc.rpm',
  }

  # Workaround for bash/rpm incompatiblity SST-2563 should fix this
  file { '/bin/bash':
    ensure => link,
    target => 'bash_64',
  }

  # AIX Package Repo - This content is not specifically profile material
  if $repo_mount {
    # Create mountpoint
    file { $repo_path:
      ensure => directory,
      before => Mount[$repo_path],
    }

    # Create filesystem mount reference (do not mount)
    mount { $repo_path:
      ensure  => 'unmounted',
      atboot  => false,
      device  => $repo_mount,
      fstype  => 'nfs',
      options => 'ro,fg,intr'
    }

    # List dependencies on this repo_path
    $pkg_repo_dependencies = [
      Exec['install-aixldap-packages-all-at-once'],
      Package['bash']
      #Package['rpm.rte'],
    ]

    # Make sure the dependencies process before the mountpoint is unmounted again.
    # NOTE: This will probably not work during the "first run"
    $pkg_repo_dependencies.each | $res | {
      # This may seem backwards, but remember the "Mount[$repo_path]" will actually unmount
      $res -> Mount[$repo_path]
    }

    # This will *mount* the pkg_repo before changing the dependent resources
    transition { "mount ${repo_path}":
      resource   => Mount[$repo_path],
      attributes => { ensure => 'mounted' },
      prior_to   => $pkg_repo_dependencies,
    }
  }

  Autofs::Mount['share'] -> Package <| title != 'bos.net.nfs.client' |>
}

```
error:
```
NoMethodError: undefined method `next_token_of' for nil:NilClass
/Users/USER/.rvm/gems/ruby-2.4.3/gems/puppet-lint-2.3.6/lib/puppet-lint/plugins/check_resources/ensure_first_param.rb:32:in `fix'
/Users/USER/.rvm/gems/ruby-2.4.3/gems/puppet-lint-2.3.6/lib/puppet-lint/checkplugin.rb:42:in `block in fix_problems'
/Users/USER/.rvm/gems/ruby-2.4.3/gems/puppet-lint-2.3.6/lib/puppet-lint/checkplugin.rb:38:in `each'
/Users/USER/.rvm/gems/ruby-2.4.3/gems/puppet-lint-2.3.6/lib/puppet-lint/checkplugin.rb:38:in `fix_problems'
/Users/USER/.rvm/gems/ruby-2.4.3/gems/puppet-lint-2.3.6/lib/puppet-lint/checks.rb:67:in `block in run'
/Users/USER/.rvm/gems/ruby-2.4.3/gems/puppet-lint-2.3.6/lib/puppet-lint/checks.rb:65:in `each'
/Users/USER/.rvm/gems/ruby-2.4.3/gems/puppet-lint-2.3.6/lib/puppet-lint/checks.rb:65:in `run'
/Users/USER/.rvm/gems/ruby-2.4.3/gems/puppet-lint-2.3.6/lib/puppet-lint.rb:205:in `run'
/Users/USER/.rvm/gems/ruby-2.4.3/gems/puppet-lint-2.3.6/lib/puppet-lint/bin.rb:66:in `block in run'
/Users/USER/.rvm/gems/ruby-2.4.3/gems/puppet-lint-2.3.6/lib/puppet-lint/bin.rb:62:in `each'
/Users/USER/.rvm/gems/ruby-2.4.3/gems/puppet-lint-2.3.6/lib/puppet-lint/bin.rb:62:in `run'
/Users/USER/.rvm/gems/ruby-2.4.3/gems/puppet-lint-2.3.6/bin/puppet-lint:7:in `<top (required)>'
/Users/USER/.rvm/gems/ruby-2.4.3/bin/puppet-lint:23:in `load'
/Users/USER/.rvm/gems/ruby-2.4.3/bin/puppet-lint:23:in `<main>'
/Users/USER/.rvm/gems/ruby-2.4.3/bin/ruby_executable_hooks:15:in `eval'
/Users/USER/.rvm/gems/ruby-2.4.3/bin/ruby_executable_hooks:15:in `<main>'
```

**Comment by TJM** [comment]
I think @abegosum and mine are the same problem, but the first one might be different.

**Comment by usev6** [comment]
As a status update: On HEAD none of the three class definitions die with an internal error:

For the first one a syntax error is diagnosed (the call to ```epp``` has no closing paren). The internal error was fixed with 8a9091a998.
```
ERROR: Syntax error on line 35
Try running `puppet parser validate <file>`
```

For the second one, the problem is fixed correctly:
```
FIXED: arrow should be on the right operand's line on line 60
```

For the third one --fix doesn't have any effect, but a warning is printed. There is at least one other  issue about a similiar problem, cmp. this comment: https://github.com/rodjek/puppet-lint/issues/410#issuecomment-406171882
```
WARNING: ensure found on line but it's not the first attribute on line 54
``` 

Maybe this issue should be closed?

**Comment by rodjek** [comment]
The problem reported in #410 has now been resolved, so all the problems reported here have been resolved.
